spec.html

```diff
@@ -13799,6 +13892,74 @@ <h1>Runtime Semantics: ArgumentListEvaluation</h1>
       </emu-clause>
     </emu-clause>
 
+    <emu-clause id="sec-optional-chains">
+      <h1>Optional Chains</h1>
+      <emu-note>An optional chain is a chain of one or more property accesses and function calls, the first of which begins with the token `?.`.</emu-note>
+      <emu-clause id="sec-optional-chaining-evaluation">
+        <h1>Runtime Semantics: Evaluation</h1>
+        <emu-grammar>
+          OptionalExpression :
+            MemberExpression OptionalChain
+            CallExpression OptionalChain
+            OptionalExpression OptionalChain
+        </emu-grammar>
+        <emu-alg>
+          1. Let _baseExpression_ be the first child of this production (i.e., this |MemberExpression|, |CallExpression|, or |OptionalExpression|).
+          1. Let _baseReference_ be ? _baseExpression_.Evaluation().
+          1. Let _baseValue_ be ? GetValue(_baseReference_).
+          1. If _baseValue_ is *undefined* or *null*, then
+            1. Return *undefined*.
+          1. Let _optionalChain_ be this |OptionalChain|.
+          1. Return ? _optionalChain_.ChainEvaluation(_baseValue_, _baseReference_).
```

Comment:
I think this is less clear than just having three different `Runtime Semantics: Evaluation ` sections, one for each of the three kinds of chain, each with their own copy of the steps preceding this line.

Reply:
Because I spent a lot of time misunderstanding this, I should probably clarify before I do any more work 😄

Basically the line that you actually take issue with is this, right?

```
1. Let _baseExpression_ be the first child of this production (i.e., this |MemberExpression|, |CallExpression|, or |OptionalExpression|).
```

It feels repetitive to have 3 different sections, but I can make the change.

Reply:
I do also take some issue with that line, but actually it's the method-style invocation of `ChainEvaluation` on the commented line which I find the least clear. The spec does allow for such invocations in some circumstances, but they're used almost exclusively for working with environment records, not to dispatch abstract operations on parse nodes.

I agree it is a tiny bit repetitive, but I think the added clarity is worth it. And it's not actually that much code which gets repeated.

Reply:
I also found this to be a bit weird. +1 to separate runtime semantics.

Reply:
I agree with @bakkot here. In general, I think it's best to stick to the "syntax-directed evaluation" style as much as possible.

Reply:
To be explicit, what I think would be the most clear would be to have three subsections under the current `Runtime Semantics: Evaluation`, each in the mold of

```
        <emu-grammar>
          OptionalExpression :
            MemberExpression OptionalChain
        </emu-grammar>
        <emu-alg>
          1. Let _baseReference_ be the result of evaluating |MemberExpression|.
          1. Let _baseValue_ be ? GetValue(_baseReference_).
          1. If _baseValue_ is *undefined* or *null*, then
            1. Return *undefined*.
          1. Return the result of performing ChainEvaluation for |OptionalChain| with arguments _baseValue_ and _baseReference_.
        <emu-alg>
```

(And to similarly rewrite the uses of `ChainEvaluation` in the `Runtime Semantics: ChainEvaluation` section to use this style of dispatch.)

Reply:
> it's the method-style invocation of `ChainEvaluation` on the commented line which I find the least clear. The spec does allow for such invocations in some circumstances, but they're used almost exclusively for working with environment records,

And module records, and internal methods.

> not to dispatch abstract operations on parse nodes.

But see PR #955.